web_poet/utils.py

```diff
@@ -126,6 +133,63 @@ def new_method(self, *args, **kwargs):
     return new_method
 
 
+def cached_method(method):
+    """A decorator to cache method or coroutine method results,
+    so that if it's called multiple times for the same instance,
+    computation is only done once.
+
+    The cache is unbound, but it's tied to the instance lifetime.
+
+    .. note::
+
+        :func:`cached_method` is needed because :func:`functools.lru_cache`
+        doesn't work well on methods: self is used as a cache key,
+        so a reference to an instance is kept in the cache, and this
+        prevents deallocation of instances.
+    """
+    name_hash = hashlib.sha1(method.__name__.encode("utf8")).hexdigest()[:4]
+    cached_meth_name = f"_cached_{method.__name__}_{name_hash}"
+    if inspect.iscoroutinefunction(method):
+        meth = _cached_method_async(method, cached_meth_name)
+    else:
+        meth = _cached_method_sync(method, cached_meth_name)
+
+    meth.cached_method_name = cached_meth_name
+    return meth
+
+
+def _cached_method_sync(method, cached_method_name, maxsize=None):
```

Comment:
It seems pointless to implement `maxsize` in these private functions if we are not exposing it through the public functions that use them.

Reply:
Right; the plan was to expose them, but I was a bit lazy :) I'll either remove it, or implement it.